### PR TITLE
Add postgres as database engine

### PR DESCRIPTION
**Pull request by tomamplius**
Allow use postgres as db engine

### CODE REVIEW

**Review by sissbruecker — COMMENTED** [review]
Thanks I'm more open to this now, as the code doesn't use any SQLite specific features anymore. For some history, also see https://github.com/sissbruecker/linkding/issues/218, and this fork: https://github.com/RoGryza/linkding-postgres.

The setup in the PR currently doesn't work, as the `psycopg2` database driver is missing. So the main issue that needs to be solved is to get the driver set up for local development and in the Docker container for production.

File: siteroot/settings/base.py [hunk 1]
```diff
@@ -204,3 +195,32 @@
...
+LD_DB_ENGINE = os.getenv('LD_DB_ENGINE', 'sqlite')
+LD_DB_HOST = os.getenv('LD_DB_HOST', 'localhost')
+LD_DB_DATABASE = os.getenv('LD_DB_DATABASE', 'linkding')
+LD_DB_LOGIN = os.getenv('LD_DB_LOGIN', 'linkding')
+LD_DB_PASSWORD = os.getenv('LD_DB_PASSWORD', '')
+LD_DB_PORT = os.getenv('LD_DB_PORT', '')
+
+if LD_DB_ENGINE == "postgres" :
+    default_database = {
+        'ENGINE': 'django.db.backends.postgresql_psycopg2',
```
[the remark this answers is not in the document]
Comment (tomamplius) [reply siteroot/settings/base.py:210]: Yes, i will add psycopg2-binary.
On my side i have use python3-psycopg2 from Debian

File: siteroot/settings/base.py [hunk 2]
```diff
@@ -204,3 +195,32 @@
 if trusted_origins:
     CSRF_TRUSTED_ORIGINS = trusted_origins.split(',')
 
+# Database
+# https://docs.djangoproject.com/en/2.2/ref/settings/#databases
+
+LD_DB_ENGINE = os.getenv('LD_DB_ENGINE', 'sqlite')
+LD_DB_HOST = os.getenv('LD_DB_HOST', 'localhost')
+LD_DB_DATABASE = os.getenv('LD_DB_DATABASE', 'linkding')
+LD_DB_LOGIN = os.getenv('LD_DB_LOGIN', 'linkding')
```
[the remark this answers is not in the document]
Comment (tomamplius) [reply siteroot/settings/base.py]: agree

File: siteroot/settings/base.py [hunk 3]
```diff
@@ -204,3 +195,32 @@
...
+LD_DB_PASSWORD = os.getenv('LD_DB_PASSWORD', '')
+LD_DB_PORT = os.getenv('LD_DB_PORT', '')
```
[the remark this answers is not in the document]
Comment (tomamplius) [reply siteroot/settings/base.py]: Fixed

File: .env.sample [hunk 4]
```diff
@@ -27,3 +27,16 @@ LD_AUTH_PROXY_LOGOUT_URL=
 # List of trusted origins from which to accept POST requests
 # See docs/Options.md for more details
 LD_CSRF_TRUSTED_ORIGINS=
+
+#Database engine can be sqlite (default) or postgres
+LD_DB_ENGINE=sqlite
+#Database name (default : linkding)
+LD_DB_DATABASE=
```
[the remark this answers is not in the document]
Comment (tomamplius) [reply .env.sample:37]: Can you confirm that new version is OK?

**Comment by sissbruecker** [comment]
Polished the docs a bit. I also had to change the setup a bit to fix issues with the Docker image. The only combination I got to work (locally at least) was to use `psycopg2` instead of `psycopg2-binary`, and then also install `libpq-dev` in the Docker image. Otherwise I got errors at runtime with libpq either not being available, or being to old.

I'm still waiting for the cross-platform image to built, I'll post an update when it's ready for testing.

**Comment by sissbruecker** [comment]
I've pushed the Docker image with Postgres support for testing to: `sissbruecker/linkding:test`.

@tomamplius Would you mind giving it a try?

**Comment by tomamplius** [comment]
I sorry i not use docker. i don't know how test it?

**Comment by sissbruecker** [comment]
Out of curiosity, how do you intend to run linkding without using Docker?

If you still want to give it a try using Docker, you can follow the instruction in the README, but use the `sissbruecker/linkding:test` image instead of `sissbruecker/linkding:latest`.

Otherwise I'll try to do some manual testing myself soon.

**Comment by tomamplius** [comment]
Hi, i create script to create Debian package from git and deploy on proxmox lxc contener with apt.
My git is https://git.lgy.fr/thomas/debmaker.
I will push linkding script later when deb file will respect (FHS)[https://fr.wikipedia.org/wiki/Filesystem_Hierarchy_Standard]

**Comment by sissbruecker** [comment]
OK got it 👍

Anyway, then I'll do some testing using Docker myself.

**Comment by tomamplius** [comment]
Thank you a lot.
And great work about this project.
I will work to new PR to add djengo-opensearch after

**Comment by sissbruecker** [comment]
The easiest way is probably to export your bookmarks from the settings, and then import them again after switching the database. This will result in some minor differences (web archive snapshots will be generated again, scraped websites titles+descriptions will become the default unless an explicit title+description has been set), but those are usually acceptable.

Otherwise SQLite allows backups to SQL files, but I have my doubts if the generated syntax works out of the box with PostreSQL. Also the schema definition would have to be removed, as that is created by linkding on startup.

**Comment by tomamplius** [comment]
pg loader is a good way https://pgloader.readthedocs.io/en/latest/ref/sqlite.html

**Comment by tomamplius** [comment]
with data only option

**Pull request merged**